keps/sig-etcd/4578-server-feature-gate/README.md

```diff
@@ -0,0 +1,331 @@
+
+On the user side, how would the users know if they are using the buggy feature if they do not remember which flags were used to start the cluster? Before feature gate, they need to search through the long logs to find the flag value when the server initially started, if the log still exists. Now there is a way to find which features are enabled on a server directly, and restart the servers with the buggy feature enabled.
+
+#### Story 2
+
+A developer added a new feature, and after a few releases, they are ready to graduate the feature.
+
+Before feature gate, to avoid breaking changes, they would need to add a stable flag `--my-feature-flag` along with `--experimental-my-feature-flag`. Now there are two duplicate flags doing exactly the same thing.
+Now they need to wait till the next minor release to remove `--experimental-my-feature-flag`. If they forget, we are stuck with the two flags until someone bothers to look into verifying the two flags are identical and eventually removing `--experimental-my-feature-flag`.
+
+Now, the developer just need to change stage of the feature to the next stage - single line change.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Is feature enablement state a server level or cluster level property?
+
+There are simple features like `ExperimentalEnableDistributedTracing` (enables distributed tracing using OpenTelemetry protocol), which could work fine at the local server level.
+
+There are also features like `ExperimentalEnableLeaseCheckpoint` (enables leader to send regular checkpoints to other members to prevent reset of remaining TTL on leader change), if different server nodes have different enablement values for `ExperimentalEnableLeaseCheckpoint`, the results would be inconsistent depending on which one is the leader and confusing.
+
+There are also more critial features like changing the Apply logic of a raft message. All members have to agree on whether to use the new logic or old logic, otherwise it could break correctness.
+
+So feature gate should be able to handle both server level and cluster level features. Since cluster level feature consensus is itself a complicated problem. We will defer it to its own KEP, and only focus on the server level feature gate in this KEP.
+
+#### Is feature gate a replacement of boolean flags?
+No. 
+Feature gate is meant to facilitate the safe development of a feature and guard any code changes behind the gate. It is not a feature configuration. A feature could have many other flags as its configuration, including boolean flags. If you think the user would still need to be able to toggle its configuration on/off when the feature graduates and becomes an integral part of etcd, then you would need a boolean configuration flag in addition to the feature gate when introducing the feature.
```

Comment:
tldr; we should only consider feature flags, for knobs that we plan to make default and force enabled for etcd instances. For optional knobs we should keep a command line flags.